search.go

```diff
@@ -0,0 +1,324 @@
+	return ws, resp, err
+}
+
+// Commits searches the expression within all commits
+//
+// GitLab API docs: https://docs.gitlab.com/ee/api/search.html#scope-commits
+func (s *SearchService) Commits(query string, opt *ListOptions, options ...OptionFunc) ([]*Commit, *Response, error) {
+	var cs []*Commit
+	resp, err := s.search("commits", query, opt, &cs)
+	return cs, resp, err
+}
+
+// CommitsByGroup searches the expression within commits for the specified
+// group
+//
+// GitLab API docs: https://docs.gitlab.com/ee/api/search.html#scope-commits
+func (s *SearchService) CommitsByGroup(gid interface{}, query string, opt *ListOptions, options ...OptionFunc) ([]*Commit, *Response, error) {
+	var cs []*Commit
+	resp, err := s.searchByGroup(gid, "commits", query, opt, &cs)
+	return cs, resp, err
+}
+
+// CommitsByProject searches the expression within commits for the
+// specified project
+//
+// GitLab API docs: https://docs.gitlab.com/ee/api/search.html#scope-commits
+func (s *SearchService) CommitsByProject(pid interface{}, query string, opt *ListOptions, options ...OptionFunc) ([]*Commit, *Response, error) {
+	var cs []*Commit
+	resp, err := s.searchByProject(pid, "commits", query, opt, &cs)
+	return cs, resp, err
+}
+
+type Blob struct {
+	Basename  string `json:"basename"`
+	Data      string `json:"data"`
+	Filename  string `json:"filename"`
+	ID        int    `json:"id"`
+	Ref       string `json:"ref"`
+	Startline int    `json:"startline"`
+	ProjectID int    `json:"project_id"`
+}
+
+// Blobs searches the expression within all blobs
+//
+// GitLab API docs: https://docs.gitlab.com/ee/api/search.html#scope-blobs
+func (s *SearchService) Blobs(query string, opt *ListOptions, options ...OptionFunc) ([]*Blob, *Response, error) {
+	var bs []*Blob
+	resp, err := s.search("blobs", query, opt, &bs)
+	return bs, resp, err
+}
+
+// BlobsByGroup searches the expression within blobs for the specified
+// group
+//
+// GitLab API docs: https://docs.gitlab.com/ee/api/search.html#scope-blobs
+func (s *SearchService) BlobsByGroup(gid interface{}, query string, opt *ListOptions, options ...OptionFunc) ([]*Blob, *Response, error) {
+	var bs []*Blob
+	resp, err := s.searchByGroup(gid, "blobs", query, opt, &bs)
+	return bs, resp, err
+}
+
+// BlobsByProject searches the expression within blobs for the specified
+// project
+//
+// GitLab API docs: https://docs.gitlab.com/ee/api/search.html#scope-blobs
+func (s *SearchService) BlobsByProject(pid interface{}, query string, opt *ListOptions, options ...OptionFunc) ([]*Blob, *Response, error) {
+	var bs []*Blob
+	resp, err := s.searchByProject(pid, "blobs", query, opt, &bs)
+	return bs, resp, err
+}
+
+func (s *SearchService) searchByProject(pid interface{}, scope, query string, opt *ListOptions, result interface{}, options ...OptionFunc) (*Response, error) {
+	project, err := parseID(pid)
+	if err != nil {
+		return nil, err
+	}
+
+	params := make(url.Values)
+	params.Set("scope", scope)
+	params.Set("search", query)
```

Comment:
Believe we should be creating an options struct for these `params.Set` (here and in the other funcs) and embedding `ListOptions` if appropriate. As I understand the options are encoded into the query when calling `NewRequest`
Ref: https://github.com/xanzy/go-gitlab/blob/master/gitlab.go#L428

Reply:
I didn't want the user to create a struct when you only want to search for a string. I wanted to make it easy for the user. I looked up go-github, they do it for the query 

https://github.com/google/go-github/blob/master/github/search.go#L186

I thought it was a pretty good idea... but I wish I could get rid of the `Set` for the `scope` parameter.

Reply:
I could hardcode the `scope` parameter in the url, but then I would've to do the params thingy for the query as github does it. I think would be not so bad.

Reply:
@svanharmelen any idea?

Reply:
Let me get back on this early next week! Sorry for the delay with this one...